File: test/scripts/07-exceptions-control-flow/correct.py
#!/usr/bin/env python3
"""
CORRECT: Check first, no exception for expected case.
Resume: "Avoid Exceptions for Control Flow"
"""
import os
import signal
import time
N = 100_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False
DATA = ["42", "invalid", "17", "bad", "99"] * (N // 5)

def parse_good(s):
    if s.isdigit() or (s.startswith('-') and s[1:].isdigit()):
        return int(s)
    return None

def main():
    global stop
    print("CORRECT: isdigit() check -> no exception")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        total = 0
        for s in DATA:
            v = parse_good(s)
            if v is not None:
                total += v
        laps += 1
    print(f"Stopped: {laps} laps")

if __name__ == "__main__":
    main()

File: test/scripts/07-exceptions-control-flow/correct.js
const N = 100_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const DATA = Array(N).fill(null).map((_, i) => ["42","invalid","17","bad","99"][i % 5]);
function parseGood(s) {
  if (/^-?\d+$/.test(s)) return parseInt(s, 10);
  return null;
}
let laps = 0;
const start = Date.now();
console.log('CORRECT: regex check -> no exception');
const id = setInterval(() => {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    clearInterval(id);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  let total = 0;
  for (let i = 0; i < DATA.length; i++) {
    const v = parseGood(DATA[i]);
    if (v !== null) total += v;
  }
  laps++;
}, 0);
process.on('SIGINT', () => { clearInterval(id); process.exit(0); });
